builtins: implement ST_RemoveRepeatedPoints

erikgrinaker opened the pull request:
Release justification: low risk, high benefit changes to existing functionality

Release note (sql change): Implement the geometry builtin
`ST_RemoveRepeatedPoints`.

Closes #49017.

erikgrinaker committed:
builtins: implement ST_RemoveRepeatedPoints

Release justification: low risk, high benefit changes to existing functionality

Release note (sql change): Implement the geometry builtin
`ST_RemoveRepeatedPoints`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53999)
<!-- Reviewable:end -->

otan commented:
This is interesting. I had to do something similar in https://github.com/cockroachdb/cockroach/pull/53970 but with the caveat that it seems to always keep the first coordinates:

```
otan=# select st_astext(st_removerepeatedpoints('linestring(1 2, 1 2)'::geometry));
      st_astext
---------------------
 LINESTRING(1 2,1 2)
(1 row)

otan=# select st_astext(st_removerepeatedpoints('linestring(1 2, 1 2, 3 4)'::geometry));
      st_astext
---------------------
 LINESTRING(1 2,3 4)
(1 row)

otan=# select st_astext(st_removerepeatedpoints('linestring(1 2, 1 2, 1 2, 3 4)'::geometry));
      st_astext
---------------------
 LINESTRING(1 2,3 4)
(1 row)

otan=# select st_astext(st_removerepeatedpoints('linestring(1 2, 1 2, 1 2)'::geometry));
      st_astext
---------------------
 LINESTRING(1 2,1 2)
(1 row)
```

can you verify that if it's all the same coordinates behaviors match too?

otan commented on pkg/geo/geomfn/remove_repeated_points_test.go:
```diff
@@ -0,0 +1,102 @@
+	}
+
+	for _, tc := range testCasesTolerance {
+		t.Run(tc.wkt, func(t *testing.T) {
```
mind appending the tolerance to the tc description?

erikgrinaker commented:
> st_removerepeatedpoints seems to always keep the first two coordinates

Good catch, thanks - these cases errored, it needs to enforce all type invariants (e.g. that linestrings have at least 2 coords, and that polygon linearrings have at least 4 coords and are closed). I've fixed the cases I could think of and added tests for them.

otan reviewed (approved):
neat, thanks!

otan commented:
bors r+
